rustls/tests/api.rs

```diff
@@ -5034,6 +5097,259 @@ fn test_server_rejects_clients_without_any_kx_group_overlap() {
     );
 }
 
+#[cfg(feature = "tls12")]
+#[test]
+fn test_server_picks_dhe_group_when_clienthello_has_no_dhe_group_in_groups_ext() {
+    fn clear_named_groups_ext(msg: &mut Message) -> Altered {
+        if let MessagePayload::Handshake { parsed, encoded } = &mut msg.payload {
+            if let HandshakePayload::ClientHello(ch) = &mut parsed.payload {
+                for mut ext in ch.extensions.iter_mut() {
+                    if let ClientExtension::NamedGroups(ngs) = &mut ext {
+                        ngs.clear();
+                    }
+                }
+            }
+            *encoded = Payload::new(parsed.get_encoding());
+        }
+        Altered::InPlace
+    }
+
+    let client_config = finish_client_config(
+        KeyType::Rsa,
+        rustls::ClientConfig::builder_with_provider(ffdhe::ffdhe_provider().into())
+            .with_protocol_versions(&[&rustls::version::TLS12])
+            .unwrap(),
+    );
+    let server_config = finish_server_config(
+        KeyType::Rsa,
+        rustls::ServerConfig::builder_with_provider(ffdhe::ffdhe_provider().into())
+            .with_protocol_versions(&[&rustls::version::TLS12])
+            .unwrap(),
+    );
+
+    let (client, server) = make_pair_for_configs(client_config, server_config);
+    let (mut client, mut server) = (client.into(), server.into());
+    transfer_altered(&mut client, clear_named_groups_ext, &mut server);
+    assert!(server.process_new_packets().is_ok());
+}
+
+#[cfg(feature = "tls12")]
+#[test]
+fn test_server_picks_dhe_group_when_clienthello_has_no_groups_ext() {
+    fn remove_named_groups_ext(msg: &mut Message) -> Altered {
+        if let MessagePayload::Handshake { parsed, encoded } = &mut msg.payload {
+            if let HandshakePayload::ClientHello(ch) = &mut parsed.payload {
+                ch.extensions
+                    .retain(|ext| !matches!(ext, ClientExtension::NamedGroups(_)));
+            }
+            *encoded = Payload::new(parsed.get_encoding());
+        }
+        Altered::InPlace
+    }
+
+    let client_config = finish_client_config(
+        KeyType::Rsa,
+        rustls::ClientConfig::builder_with_provider(ffdhe::ffdhe_provider().into())
+            .with_protocol_versions(&[&rustls::version::TLS12])
```

Comment:
Nitpick: it's better to use `.with_safe_default_protocol_versions()` to keep sync with other tests.

Please be free to resolve this comment, if this test case needs TLS 1.2.

Reply:
I think the client needs to be TLS 1.2 only. I can configure server with `.with_safe_default_protocol_versions()` instead.